### ISSUE TITLE
`SgdConfig` has bad documentation and a bad signature for `init`

### ISSUE THREAD

**Issue by miestrode** [burn-rs/burn#868]
Currently the documentation for the `init` method for `SgdConfig` states:

> Creates a new [SgdConfig](https://docs.rs/burn/latest/burn/optim/struct.SgdConfig.html) with default values.

This is wrong, as `init` methods of configs should create the thing they are configurating. Indeed this is what `init` for `SgdConfig` does.

Additionally the return type for the method is `OptimizerAdaptor<Sgd<...>, M, B>`, even though this may change in the future if `Sgd` were to not need an adaptor, or if optimizer adaptors were ever changed or scrapped. shouldn't ideally the return type be an `impl Optimizer`?

**Comment by nathanielsimard** [comment]
Yes and yes!!

**Comment by miestrode** [comment]
Looking more at the codebase though, do you think it might be better to open a new issue that is more relevant for this? I think the whole `optim` module really desperately needs a refactor in some areas, because not only `Sgd` has these issues, it's just the optimizer I used at the time, which made me notice this issue. The documentation all around doesn't feel consistent at all, kind of like several different people with different standards were writing them (which is likely the case).

It also looks like configuration structs like `MomentumConfig` don't use the approach of using `init` for initializing the target struct, and instead bundle with the target struct a `new` method. Should this be done? Is this an artifact of an older version of the codebase, or is this something else?

Finally, changing the signature to `impl Optimizer` is obviously not backwards compatible, so it might be useful to check if anyone actively depends on the current signature, and possibly push this change further ahead.

**Comment by nathanielsimard** [comment]
Yes I think we could refactor a bit more this part of the codebase! Breaking changes are OK at this point, we should strive for improving the UX and the overall framework at the cost of small breaking changes. Since we have many examples, it's quite easy to see how hard it is to update to a new burn version, and I expect this signature update to be invisible if you use `burn-train`.

**Comment by anthonytorlucci** [comment]
@miestrode @nathanielsimard, has there been any movement here? Is there a plan to refactor the `optim` module? I don't have the capacity to take on the full refactor and possibly redesign, but if there is a task within the epic, please let me know. I could try and work on it.

**Comment by nathanielsimard** [comment]
No active plan right now, but not against starting one. We're really busy with performance optimizations, but yeah this could be improved. I think starting with a PR that uses `impl Optimizer` instead of generic is a good start.
